base/multidimensional.jl

```diff
@@ -178,100 +172,164 @@ end  # IteratorsMD
+# Both fast
+stagedfunction _unsafe_getindex!(::LinearFast, dest::AbstractArray, ::LinearFast, src::AbstractArray, I::Union(Real, AbstractVector, Colon)...)
     N = length(I)
     Isplat = Expr[:(I[$d]) for d = 1:N]
     quote
-        checksize(dest, $(Isplat...))
+        stride_1 = 1
+        @nexprs $N d->(stride_{d+1} = stride_d*size(src, d))
+        $(symbol(:offset_, N)) = 1
         k = 1
-        @nloops $N i dest d->(@inbounds j_d = unsafe_getindex(I[d], i_d)) begin
-            @inbounds dest[k] = (@nref $N src j)
+        @nloops $N i dest d->(@inbounds offset_{d-1} = offset_d + (unsafe_getindex(I[d], i_d)-1)*stride_d) begin
+            unsafe_setindex!(dest, unsafe_getindex(src, offset_0), k)
             k += 1
         end
         dest
     end
 end
-
-# Version that uses linear indexing for src
-stagedfunction _getindex!(dest::Array, src::Array, I::Union(Int,AbstractVector)...)
+# Fast destination, slow source
+stagedfunction _unsafe_getindex!(::LinearFast, dest::AbstractArray, ::LinearSlow, src::AbstractArray, I::Union(Real, AbstractVector, Colon)...)
     N = length(I)
     Isplat = Expr[:(I[$d]) for d = 1:N]
     quote
-        checksize(dest, $(Isplat...))
-        stride_1 = 1
-        @nexprs $N d->(stride_{d+1} = stride_d*size(src,d))
-        @nexprs $N d->(offset_d = 1)  # only really need offset_$N = 1
         k = 1
-        @nloops $N i dest d->(@inbounds offset_{d-1} = offset_d + (unsafe_getindex(I[d], i_d)-1)*stride_d) begin
-            @inbounds dest[k] = src[offset_0]
+        @nloops $N i dest d->(@inbounds j_d = unsafe_getindex(I[d], i_d)) begin
+            v = @ncall $N unsafe_getindex src j
+            unsafe_setindex!(dest, v, k)
             k += 1
         end
         dest
     end
 end
-
-# It's most efficient to call checkbounds first, then to_index, and finally
-# allocate the output. Hence the different variants.
-_getindex(A, I::(Union(Int,AbstractVector)...)) =
-    _getindex!(similar(A, index_shape(I...)), A, I...)
-
-# The stagedfunction here is just to work around the performance hit
-# of splatting
-stagedfunction getindex(A::Array, I::Union(Real,AbstractVector)...)
+# A slow destination. It's unlikely a fast array would give a slow similar array
+# so it's not worth specializing that case.
+stagedfunction _unsafe_getindex!(::LinearSlow, dest::AbstractArray, ::LinearIndexing, src::AbstractArray, I::Union(Real, AbstractVector, Colon)...)
     N = length(I)
     Isplat = Expr[:(I[$d]) for d = 1:N]
     quote
-        checkbounds(A, $(Isplat...))
-        _getindex(A, to_index($(Isplat...)))
+        @nloops $N i dest d->(@inbounds j_d = unsafe_getindex(I[d], i_d)) begin
+            v = @ncall $N unsafe_getindex src j
+            @ncall $N unsafe_setindex! dest v i
+        end
+        dest
     end
 end
 
-# Also a safe version of getindex!
-stagedfunction getindex!(dest, src, I::Union(Real,AbstractVector)...)
+checksize(A::AbstractArray, I::AbstractArray) = true # reshaped linear idx
```

Comment:
Ah, good catch.  I think this is leftover from an attempted refactor of `checksize` that didn't use stagedfunctions.  It's incorrect in either case, though.